src/components/download/download.ts

```diff
@@ -35,9 +45,29 @@ export async function toTempFile(sourceUrl: string): Promise<Errorable<string>>
 export async function to(sourceUrl: string, destinationFile: string): Promise<Errorable<null>> {
     ensureDownloadFunc();
     try {
-        await download(sourceUrl, path.dirname(destinationFile), { filename: path.basename(destinationFile) });
+        await download!(sourceUrl, path.dirname(destinationFile), { filename: path.basename(destinationFile) });  // safe because we ensured it
         return { succeeded: true, result: null };
     } catch (e) {
         return { succeeded: false, error: [e.message] };
     }
 }
+
+export async function once(sourceUrl: string, destinationFile: string): Promise<Errorable<null>> {
+    const downloadStatus = DOWNLOAD_ONCE_STATUS[destinationFile];
+    if (!downloadStatus || downloadStatus === DownloadOperationStatus.Failed) {
+        DOWNLOAD_ONCE_STATUS[destinationFile] = DownloadOperationStatus.Queued;
+        const result = await to(sourceUrl, destinationFile);
+        DOWNLOAD_ONCE_STATUS[destinationFile] = succeeded(result) ? DownloadOperationStatus.Completed : DownloadOperationStatus.Failed;
+        return result;
+    } else {
+        while (true) {
```

Comment:
I feel like this could have been done with an event handler somwhere instead of this while loop?

Reply:
I'm not sure.  It needs to be something `await`able unless I rework the whole thing into callback style.  I guess the map could contain a promise which the `else` clause could `await`... but it feels like it could get fiddlier than the loop is... I agree the loop does not spark joy though...